jupyter_client/tests/signalkernel.py

```diff
@@ -0,0 +1,57 @@
+"""Test kernel for signalling subprocesses"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+from __future__ import print_function
+
+from subprocess import Popen, PIPE
+import sys
+import time
+
+from ipykernel.kernelbase import Kernel
+from ipykernel.kernelapp import IPKernelApp
+
+
+class SignalTestKernel(Kernel):
+    """Kernel for testing subprocess signaling"""
+    implementation = 'signaltest'
+    implementation_version = '0.0'
+    banner = ''
+    
+    def __init__(self, **kwargs):
+        super(SignalTestKernel, self).__init__(**kwargs)
+        self.children = []
+        
+
+    def do_execute(self, code, silent, store_history=True, user_expressions=None,
+                   allow_stdin=False):
+        code = code.strip()
+        reply = {'status': 'ok'}
+        if code == 'start':
+            child = Popen(['bash', '-i', '-c', 'sleep 30'], stderr=PIPE)
+            self.children.append(child)
+            reply['pid'] = self.children[-1].pid
```

Comment:
We're sticking arbitrary extra fields in the reply? I'm a bit surprised none of our machinery complains about that. Do we want to assume that will always be the case? We could stick them in metadata or `user_expressions`.